Cleaner errors and optionally enable skipping wrong packages

htgoebel opened the pull request:
This brings in https://github.com/dkg/python-pgpdump/tree/cleaner-errors plus some minor changes by me.

dkg committed:
pass on unknown pubkey algorithms rather than raising an Exception

If we're OK with "passing" on experimental algorithms, we should be ok
with passing on known algorithms that we know the name of but don't
implement.

dkg committed:
Correct copy/pasted error message

dkg committed:
rfc4880bis includes embedded issuer fingerprint

The issuer fingerprint effectively supercedes issuer keyID as a more
robust signal of the issuer (though it's possible and in some cases
likely to include both in a signature, to ensure that it's understood
by old and new code).

dkg committed:
Raise PgpdumpException when MPI parsing fails

Some malformed MPIs will cause a failure in packet parsing; this is an
OpenPGP formatting error, so we should raise it explicitly.

Below is an example packet (found on the SKS keyserver network) that
has a malformed MPI.

-----BEGIN PGP PUBLIC KEY BLOCK-----

mQGiBEhFwD0RBAC8mLriGM4hMjV4ePeBD+FoqB/ENCXbLgl1vAvB3ha/HxqVeH1o
w8HdYiSHykreflEqH31fxJv1xFgQ87sf21jv/8TeR0LeXnx7eEIF+uhDHWiZIr4q
+EfrzIkpC9+yyo5XaveqNqiSOcnBH2aq334L80/2LR++7J2ffZHgWVfKDtsofE6L
XfmOPAwT7R06+po2VnFjlwsD/jXgG6rkjwJZn/VhibcifIms66Fp/RYCXZp9mCeS
pADN7i+7VwwCjaME3ndQu/KMdMqSzfHsWJFJbycjN5MK49vxXZyz+cGblcKmZH3V
CKwC2iVVRP5FJe+ZLaqmEt/qbyjstXimWUYaJkQkD3lPz+7NfVhIXa0tlE2UcL+7
A5ghA/9+GZu0eFtV7OokUuyPHcfQDQWLh+nuJP7bVSepo9Gy4c0SCYyKlHf+e+A3
qIF76+1Gj7IvVEY8fcC0APzGad9Y0WELSTNcZ4+UqW3Ezag1k7Yq5F0VycZ2LTIC
9nF3UsjKgUxWehbkrNz9NRx9v7ktxS0f2N43dEQIffab3rPqMw==
=V948
-----END PGP PUBLIC KEY BLOCK-----

Signed-off-by: Daniel Kahn Gillmor <dkg@fifthhorseman.net>

fix up mpi error

Signed-off-by: Daniel Kahn Gillmor <dkg@fifthhorseman.net>

dkg committed:
add optional skip= parameter to data.BinaryData.packets()

Each malformed packet skipped will generate a warning with the python
logging module, but the stream will continue to return any subsequent
well-formed packets as it is encountered.

Signed-off-by: Daniel Kahn Gillmor <dkg@fifthhorseman.net>

dkg committed:
raise a PgpdumpException on malformed User Attribute packets

dkg committed:
raise errors for wrong-size ints

dkg committed:
add encode_packet, and dump encoded packet to log when malformed parsing

dkg committed:
catch errors in new_tag_length

Signed-off-by: Daniel Kahn Gillmor <dkg@fifthhorseman.net>

dkg committed:
throw a PgpdumpException on malformed image attributes

Signed-off-by: Daniel Kahn Gillmor <dkg@fifthhorseman.net>

htgoebel committed:
Cleanup: Reformat some code.

to make it more PEP 8 compliant.

htgoebel committed:
Merge branch 'master' into cleaner-errors.

htgoebel committed:
Cleanup: Break some very long lines.

htgoebel commented:
Please note: This pull-request also includs some changes already in #3 due to how dkg did apply the changes in his repo. Anyway, please merge #2 first to have that branch merge into here. Thus others can easily see that dpk'g changes are in here already. Thanks.

alex-nitrokey commented:
Thank you very much for your effort creating this PR! I will look into this one during the week.

alex-nitrokey commented on pgpdump/utils.py:
```diff
@@ -71,27 +73,93 @@ def crc24(data):
     return (get_int4(data, offset) << 32) + get_int4(data, offset + 4)
 
 
+def encode_packet(tag, new, data, armored=False):
```
It would be great to have a function description here.

alex-nitrokey reviewed (changes requested):
Thanks again for this PR. Would you mind creating a test case in test.py that checks and demonstrates the new feature of packet skipping and encoding? That'd be awesome. I guess, I would merge this PR anyway, but I would really prefer having this tested in future to prevent breaking stuff unintentionally.

htgoebel committed:
Add some comments.

htgoebel committed:
utils: Add doc-string to encode_packet.

htgoebel committed:
tests: Add tests for "encode_packet".

htgoebel commented:
@alex-nitrokey Sorry for the delay, I was completely distracted from this issue. Now I updates the pull-request and added what you requested.

htgoebel committed:
tests: Add tests for malformed mpi.

The check was inroduced in 25cc2738b872eb3cf48acdb193a7a21cf2478679,
but not tests provided.

htgoebel commented:
Lint errors are from code untouched by me.

techge committed:
Fix lint errors

alex-nitrokey commented:
> Sorry for the delay, I was completely distracted from this issue. Now I updates the pull-request and added what you requested.

No harm done, thanks for your contribution and efforts!

> Lint errors are from code untouched by me.

Some are ;) I think I fixed them, but couldn't apply them directly to you PR, so I created a [new branch](https://github.com/alex-nitrokey/python-pgpdump/tree/lint-fixes)... See [this commit](https://github.com/alex-nitrokey/python-pgpdump/commit/b3b858a9fde21d2443e02401e51cbadec9b4572c). If you check the correct box in the PR settings I might be able to include it right away here. Or please see if you can just merge the commit. Or I will just merge your PR and fix the minor lint errors afterwards.

Anyway I thank you for contributing to this software I am not even using anymore...

htgoebel committed:
Fix lint errors.

alex-nitrokey commented:
Again, thanks a lot!
